### DISCUSSION TITLE
Creating a custom Attached Property and setting its value in XAML

### DISCUSSION THREAD

**Discussion by npolyak** [AvaloniaUI/Avalonia#6123] [Q&A] [answered]
I am trying to create a custom attached property TheWidth: and then refer to it in XAML. Here is my XAML code:

    <Grid local:AttachedProperties.TheWidth="250">

And here is my attached property code:

    public class AttachedProperties : AvaloniaObject
    {
         public static readonly AttachedProperty<double> TheWidthProperty =
               AvaloniaProperty.RegisterAttached<AttachedProperties, Interactive, double>
               (
               "TheWidth",
               default(double),
               false);
     }

The compiler complains about it "Unable to resolve suitable regular or attached property TheWidth on type SimpleAvalonTests:SimpleAvalonTests.AttachedProperties

What am I doing wrong?

**Comment by maxkatz6** [comment]
You are missing generic parameters.
Let's take an example of Tooltip.Tip:
```csharp
public static readonly AttachedProperty<object?> TipProperty =
    AvaloniaProperty.RegisterAttached<ToolTip, Control, object?>("Tip");
```

Here "object?" is the type of value that this property contain.
"ToolTip" is the type of property owner, usually of the class where it will be used and where it was defined.
 "Control" is the type of property host, on which you can set or get this property.

See definition of RegisterAttached method https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaProperty.cs#L267-L287

**Reply by maxkatz6** [reply]
So in your case I believe it should be something like this:
```csharp
public class AttachedProperties : AvaloniaObject
{
    public static readonly AttachedProperty<double> TheWidthProperty =
        AvaloniaProperty.RegisterAttached<AttachedProperties, Control, double>("TheWidth");
} 
```
(default value and "inherits" are default in your example, so can be skipped)

**Reply by npolyak** [reply]
I just realized that your code is exactly the same as mine (see above - I fixed the code style), aside from using Control as a second generic argument instead of Interactive. I changed to Control and I still have the same problem. Is something wrong with my XAML? I made it to be the same as in WPF

**Comment by jp2masa** [comment, marked as answer]
I think you have to define a getter and setter like this:

```cs
public static double GetTheWidth(AvaloniaObject obj) => obj.GetValue(TheWidthProperty);

public static void SetTheWidth(AvaloniaObject obj, double value) => obj.SetValue(TheWidthProperty, value);
```

**Reply by npolyak** [reply]
Thanks JP! it worked!!!
